Expand Traits, add associated types (#1815)
google/comprehensive-rust: src/methods-and-traits/traits/implementing.md

Reviewer: ![image](https://github.com/google/comprehensive-rust/assets/65899331/b7080977-b6ea-4588-a562-701e38cea9da)

Looks good . I think we can omit this point because, I think the example above it self is explains the syntax. As, everyone learns rust after some programming language.

You can add point like from above only `Dog` struct can use this `Pet trait` no other struct can use this something like this. What do you say .

Author: That's a good point - this is different from Go "interfaces" where just implementing the right methods is enough.

Reviewer: Sounds good. okay then, You can merge the  PR. That all, I had.

Reviewer: Ohh, appologies for my previous deleted comment, I misread it.

# Implementing Traits

```rust,editable
trait Pet {
    fn talk(&self) -> String;

    fn greet(&self) {
        println!("Oh you're a cutie! What's your name? {}", self.talk());
    }
}

struct Dog {
    name: String,
    age: i8,
}

impl Pet for Dog {
    fn talk(&self) -> String {
        format!("Woof, my name is {}!", self.name)
    }
}

fn main() {
    let fido = Dog { name: String::from("Fido"), age: 5 };
    fido.greet();
}
```

<details>

> - Types must implement the trait with an `impl <trait> for <type> { .. }` block.

Reviewer 2: I'm not a fan of the mix between real syntax and pseudo-code, especially when Rust uses `<...>` for real things :smile: 

What about:

```suggestion
- To implement `Trait` for `Type`, you use an `impl Trait for Type { .. }` block.
```

- Unlike Go interfaces, just having matching methods is not enough: a `Cat` type
  with a `talk()` method would not automatically satisfy `Pet` unless it is in
  an `impl Pet` block.

- Traits may specify pre-implemented (provided) methods and methods that users
  are required to implement themselves. Provided methods can rely on required

>   methods. In this case, `greet` is provided, and relies on `talk`.

Author: TIL! I wonder where I picked up the term "provided"..

>   methods. In this case, `greet` is provided, and relies on `talk`.

Reviewer 2: Sorry that the wrapping is off (most likely).
